### PR TITLE
Lots of refactors and documentation

### PR DESCRIPTION
**Pull request by jberryman**
- Remove Joinable class with single instance
- Refactor mkQuery and produceBatch for a little more clarity, so they
  only take and return relevant data.
- Simplify `rebuildFieldStrippingRemoteRels`; there seemed to be some
  unnecessary and also insufficient filtering logic happening here (but
  could use a close review)
- factor out the remote query aliasing/sorting stuff into
  `enumerateRowAliases`, `unEnumerateRowAliases` which we can document
  properly
- Remove some unnecessary fields from Batch; narrowing scope of
  functions that take or return a Batch aids understanding.
- inline stuff into `extractRemoteRelArguments` and refactor
- add `coerceSet` to Hasura.Prelude (this was used several times in a
  previous iteration, now only appears once; I'm not super committed to
  keeping it)
- add some tests, mostly just trying to excercise a few different things
  in the same query and used as an aid to understanding the code

### CODE REVIEW

File: server/src-lib/Hasura/GraphQL/Execute/RemoteJoins.hs [hunk 1]
```diff
@@ -106,50 +133,18 @@ rebuildFieldStrippingRemoteRels =
...
--- | Get a list of fields needed from a hasura result.
-neededHasuraFields
-  :: RemoteField -> [FieldName]
-neededHasuraFields = toList . rtrHasuraFields . rmfRemoteRelationship
 
 -- | Join the data returned from the original hasura request with the remote values.
-joinResults :: GQRespValue
-            -> [(Batch, EncJSON)]
-            -> GQRespValue
+joinResults :: GQRespValue -> [(Batch, EncJSON)] -> GQRespValue
```
Comment (tirumaraiselvan) [inline server/src-lib/Hasura/GraphQL/Execute/RemoteJoins.hs]: Would have been more readable if the signature was:

`joinResults :: GQRespValue -> [(RemoteRelField, EncJSON)] -> GQRespValue`


See comment on `data Batch`

File: server/src-lib/Hasura/GraphQL/Execute/RemoteJoins.hs [hunk 2]
```diff
@@ -318,28 +309,29 @@ peelOffNestedFields topNames topValue = foldM peel topValue (NE.drop 1 topNames)
...
 
 
--- TODO document all of this
+-- TODO In what sense is this a "batch"?
 data Batch =
   Batch
-    { batchRemoteTopQuery        :: !VQ.RemoteTopField
-    , batchRelFieldPath          :: !RelFieldPath
-    , batchIndices               :: ![ArrayIndex]
+    { batchRelFieldPath          :: !RelFieldPath
```
Comment (tirumaraiselvan) [inline server/src-lib/Hasura/GraphQL/Execute/RemoteJoins.hs]: Now that `RemoteTopField` is removed from `Batch` (surprising!), I am not sure what is the point of this data type in the first place. Seems like everything is available in `RemoteRelField` no?

This was supposed to be a batch in the sense it has the query (in the form of VQ.RemoteTopField) which can then make the following query:

```
query {
  hasura_idx_1 : field(..) { .. }
  hasura_idx_2 : field(..) { .. }
  hasura_idx_3 : field(..) { .. }
  ...
}
```

File: server/src-lib/Hasura/GraphQL/Execute/RemoteJoins.hs [hunk 3]
```diff
@@ -164,54 +159,48 @@ emptyResp = GQRespValue OJ.empty VB.empty
...
-    remoteDataE = let indexedRemotes = map getIdxRemote $ OJ.toList $ _gqRespData remoteResp
-                  in case any isNothing indexedRemotes of
-                       True -> Left "couldn't parse all remote results"
-                       False -> Right $ map snd $ sortOn fst $ catMaybes indexedRemotes
-
-    getIdxRemote (remoteAlias, value) = let idxStr = stripPrefix "hasura_array_idx_" (T.unpack remoteAlias)
-                                  in (, value) <$> (join $ readMaybe <$> idxStr :: Maybe Int)
-
-    cardinality = biCardinality batchInputs
+    -- Re-sort expected aliases to account for naughty remotes that don't
```
Comment (tirumaraiselvan) [inline server/src-lib/Hasura/GraphQL/Execute/RemoteJoins.hs]: untrusted remotes can be very naughty indeed ;)

File: server/src-lib/Hasura/GraphQL/Execute/RemoteJoins.hs [hunk 4]
```diff
@@ -49,54 +53,77 @@ import           Hasura.Prelude
...
+        fmap join $ flip Seq.traverseWithIndex ss0 $
+          \idx subfield ->
              case _fRemoteRel subfield of
-               Nothing -> fmap Right (rebuild idx thisPath subfield)
+               Nothing -> Seq.singleton <$> rebuild idx thisPath subfield
                Just remoteField -> do
+
                  modify (remoteRelField :)
-                 pure (Left remoteField)
+                 -- NOTE: this may result in redundant fields in the SELECT, but this seems fine.
```
Comment (tirumaraiselvan) [inline server/src-lib/Hasura/GraphQL/Execute/RemoteJoins.hs]: Not sure I understand what you mean by `redundant fields in the SELECT`?

File: server/src-lib/Hasura/GraphQL/Execute/RemoteJoins.hs [hunk 5]
```diff
@@ -49,54 +53,77 @@ import           Hasura.Prelude
...
+--
+-- 'Nothing' means there were no remote fields.
+--
+-- Output here is eventually assembled into a 'QExecPlan' 'Tree' in
+-- 'getExecPlan', and finally executed in 'runGQ'.
+rebuildFieldStrippingRemoteRels
+  :: VQ.Field 
+  -> Maybe (VQ.Field, NonEmpty RemoteRelField)
+  -- ^ NOTE: the ordering of the _fSelSet fields in the result 'VQ.Field' seems not to matter at 
+  -- all for correctness here (based on experimentation)... I haven't puzzled out why yet.
```
Comment (tirumaraiselvan) [inline server/src-lib/Hasura/GraphQL/Execute/RemoteJoins.hs]: How are you testing the ordering of fields? The python test suite does not account for ordering so this might have failed to show there.

File: server/src-lib/Hasura/GraphQL/Execute/RemoteJoins.hs [hunk 6]
```diff
@@ -164,54 +159,48 @@ emptyResp = GQRespValue OJ.empty VB.empty
...
+    Nothing -> appendJoinError accumResp $ 
+      "did not find expected hasura alias prefex in all remote results: " <>
+      fromString (show $ _gqRespData remoteResp)
+    Just remoteData0 -> do
       case inHashmap batchRelFieldPath accumData0 remoteData0 of
-        Left err       -> appendJoinError accumRespWithRemoteErrs err
-        Right (val, _) -> set gqRespData val accumRespWithRemoteErrs
+        Left err -> appendJoinError accumRespWithRemoteErrs err
+        -- TODO assert something about _remainingRemote?
+        Right (val, _remainingRemote) -> set gqRespData val accumRespWithRemoteErrs
```
Comment (tirumaraiselvan) [inline server/src-lib/Hasura/GraphQL/Execute/RemoteJoins.hs]: Ideally `_remainingRemotes` should be an empty list since we expect all the remotes to have been joined.

File: server/src-lib/Hasura/GraphQL/Execute/RemoteJoins.hs [hunk 7]
```diff
@@ -164,54 +159,48 @@ emptyResp = GQRespValue OJ.empty VB.empty
 -- | Insert at path, index the value in the larger structure.
```
Comment (tirumaraiselvan) [inline server/src-lib/Hasura/GraphQL/Execute/RemoteJoins.hs]: We should make sure that `insertBatchResults` is not slower than O(n) algorithm to begin with, where `n` is number of remote results. This can just be added as a comment and then improved in the later PR.

File: server/src-lib/Hasura/GraphQL/Execute/RemoteJoins.hs [hunk 8]
```diff
@@ -283,14 +272,16 @@ insertBatchResults accumResp Batch{..} remoteResp =
...
     -- at the right index), removing phantom fields
+    spliceRemote :: OJ.Value -> OJ.Object -> Either GQJoinError OJ.Object 
     spliceRemote remoteDataVal accumData = do
       peeledValue <- peelOffNestedFields batchNestedFields remoteDataVal
       pure $
-        -- TODO Brandon note:
-        -- document rrPhantomFields/batchPhantoms so that we can understand if or why it's okay that the OJ.delete might silently be a noop here:
+        -- Strip phantom fields
         foldl' (flip OJ.delete)
                 -- TODO Brandon note:
```
Comment (tirumaraiselvan) [inline server/src-lib/Hasura/GraphQL/Execute/RemoteJoins.hs]: Good points. Hopefully these will be resolved automatically while sending out the new PRs

File: server/src-lib/Hasura/GraphQL/Execute/RemoteJoins.hs [hunk 9]
```diff
@@ -348,46 +340,53 @@ produceBatch rtqOperationType rtqRemoteSchemaInfo RemoteRelField{..} batchInputs
...
 fieldCallsToField ::
      Map.HashMap G.Name AnnInpVal
   -> Map.HashMap G.Variable G.ValueConst
   -> SelSet
-  -> Maybe G.Alias
+  -> G.Alias
   -> NonEmpty FieldCall
   -> Field
-fieldCallsToField userProvidedArguments variables finalSelSet = nest
+fieldCallsToField userProvidedArguments variables finalSelSet topAlias =
```
Comment (tirumaraiselvan) [inline server/src-lib/Hasura/GraphQL/Execute/RemoteJoins.hs]: is "Call" a standard terminology? `FieldCall` seems like a bad name if it isn't derivable from standard terminology no?

File: server/src-lib/Hasura/GraphQL/Resolve/Select.hs [hunk 10]
```diff
@@ -95,13 +95,11 @@ fromAggSelSet
...
-  withSelSet selSet $ \f -> do
-    let fTy = _fType f
-        fSelSet = _fSelSet f
-    case _fName f of
+  withSelSet selSet $ \Field{..} -> 
+    case _fName of
       "__typename" -> return $ RS.TAFExp $ G.unName $ G.unNamedType fldTy
-      "aggregate"  -> RS.TAFAgg <$> convertAggFld fTy fSelSet
-      "nodes"      -> RS.TAFNodes <$> fromSelSet fTy fSelSet
+      "aggregate"  -> RS.TAFAgg <$> convertAggFld _fType _fSelSet
```
Comment (tirumaraiselvan) [inline server/src-lib/Hasura/GraphQL/Resolve/Select.hs]: is it important to change these functions? Seems unrelated to remote joins?

File: server/src-lib/Hasura/GraphQL/Transport/HTTP.hs [hunk 11]
```diff
@@ -42,31 +42,27 @@ runGQ reqId userInfo reqHdrs req = do
...
-                  (\(E.QExecPlanUnresolved remoteRelField _) -> remoteRelField)
-                  unresolvedPlans
           let (initValue, remoteBatchInputs) =
-                E.extractRemoteRelArguments
-                  initJson
-                  remoteRels
+                E.extractRemoteRelArguments initJson $
+                  map E.remoteRelField unresolvedPlans
+
+          -- TODO This zip may discard some unresolvedPlans when permissions
```
Comment (tirumaraiselvan) [inline server/src-lib/Hasura/GraphQL/Transport/HTTP.hs]: what permissions do you mean?

File: server/src-lib/Hasura/GraphQL/Transport/HTTP.hs [hunk 12]
```diff
@@ -42,31 +42,27 @@ runGQ reqId userInfo reqHdrs req = do
...
+          -- come into play. It's not totally clear to me if this is correct,
+          -- or how it works. Can use Data.Align for safer zips, but it seems like 
+          -- 'extractRemoteRelArguments' should be returning the zipped data.
+          let batchesRemotePlans =
+                -- TODO pass 'G.OperationType' properly when we support mutations, etc.
+                zipWith (E.mkQuery G.OperationTypeQuery) remoteBatchInputs unresolvedPlans
```
Comment (tirumaraiselvan) [inline server/src-lib/Hasura/GraphQL/Transport/HTTP.hs]: If we can get rid of `Batch` (see comment on `data Batch` above), then we won't have to zipWith here.

File: server/src-lib/Hasura/GraphQL/Transport/HTTP.hs [hunk 13]
```diff
@@ -120,13 +112,13 @@ runHasuraGQ reqId query userInfo resolvedOp = do
   return $ encodeGQResp $ GQSuccess resp
 
 
-getMergedGQResp :: [EncJSON] -> Either String GQRespValue
+getMergedGQResp :: Traversable t=> t EncJSON -> Either String GQRespValue
```
Comment (tirumaraiselvan) [inline server/src-lib/Hasura/GraphQL/Transport/HTTP.hs]: We need to make sure the algorithm is not slower than O(n) here where `n` is the number of remote relationships

**Comment by tirumaraiselvan** [comment]
I have kind of ignored 2 functions: `insertBatchResults` and `extractRemoteRelArguments` for now. Will get to it after the pending comments are resolved.

File: server/src-lib/Hasura/GraphQL/Validate/Field.hs [hunk 14]
```diff
@@ -49,13 +50,15 @@ type ArgsMap = Map.HashMap G.Name AnnInpVal
...
--- N.B. This is a tree via 'SelSet'
 -- | https://graphql.github.io/graphql-spec/June2018/#sec-Language.Fields 
+--
+-- N.B. This is a tree via 'SelSet'
 data Field
   = Field
   { _fAlias     :: !G.Alias
   , _fName      :: !G.Name
   , _fType      :: !G.NamedType
+  -- ^ TODO: this seems to get initialized to "unknown.."; how is this used? Document or improve.
```
Comment (tirumaraiselvan) [inline server/src-lib/Hasura/GraphQL/Validate/Field.hs]: This is initialized to `unknown_type` only in the RemoteJoins module. This is because we are constructing a nested `Field` and we don't have the "named type" of the parent fields. It's not important also as `Field` is more annotated than required for constructing a remote query.

File: server/src-lib/Hasura/GraphQL/Execute/RemoteJoins.hs (hunk 5 above)
Comment (jberryman) [reply server/src-lib/Hasura/GraphQL/Execute/RemoteJoins.hs]: Right, I did a temporary rebase on my branch here where I can assert ordering: https://github.com/hasura/graphql-engine/pull/2831  (this is one of the PRs I worked on but set aside).

File: server/src-lib/Hasura/GraphQL/Execute/RemoteJoins.hs (hunk 9 above)
Comment (jberryman) [reply server/src-lib/Hasura/GraphQL/Execute/RemoteJoins.hs]: Yeah I did think about this a little, but then gave up. I think we can come up with a better name for `fieldCallsToField` for sure. 

`FieldCall` might be a good enough name with good docs. It's a "field" without a selection set, where the spec says "Object fields are conceptually functions which yield values." (which is already confusing terminology, but should be familiar). But no "field call" isn't standard terminology afaict

File: server/src-lib/Hasura/GraphQL/Resolve/Select.hs (hunk 10 above)
Comment (jberryman) [reply server/src-lib/Hasura/GraphQL/Resolve/Select.hs]: That's to squash shadowing errors that came up when I derived lenses for `Field` unfortunately

File: server/src-lib/Hasura/GraphQL/Transport/HTTP.hs (hunk 11 above)
Comment (jberryman) [reply server/src-lib/Hasura/GraphQL/Transport/HTTP.hs]: For instance in `basic_relationship_with_permissions1.yaml`, where we filter a field to null or an empty list; the two lists have unequal length there.

File: server/src-lib/Hasura/GraphQL/Validate/Field.hs (hunk 14 above)
Comment (jberryman) [reply server/src-lib/Hasura/GraphQL/Validate/Field.hs]: Maybe I'll make a smart constructor for this where we can put docs

File: server/src-lib/Hasura/GraphQL/Execute/RemoteJoins.hs (hunk 4 above)
Comment (jberryman) [reply server/src-lib/Hasura/GraphQL/Execute/RemoteJoins.hs]: Sorry I meant to trace this again, but I saw a select with json aggregation with duplicated keys which postgres just ignores/unions in some way.

File: server/src-lib/Hasura/GraphQL/Execute/RemoteJoins.hs [hunk 15]
```diff
@@ -49,54 +53,77 @@ import           Hasura.Prelude
 import           Hasura.RQL.Types
 import           Hasura.SQL.Value
 
-newtype RemoteRelKey =
-  RemoteRelKey Int
-  deriving (Eq, Ord, Show, Hashable)
-
+-- | Like a 'VQ.Field' but with a path indicating where its response should be
+-- spliced back into the hasura result.
 data RemoteRelField =
```
[the remark this answers is not in the document]
Comment (jberryman) [reply server/src-lib/Hasura/GraphQL/Execute/RemoteJoins.hs]: I changed it to `RemoteRelBranch` since it's like part of a tree cut off, but with a path indicating where it fits back into the tree... or something like that...

File: server/src-lib/Hasura/GraphQL/Execute.hs [hunk 16]
```diff
@@ -82,25 +80,31 @@ data QExecPlanResolved
...
-      JoinParams opType rows = joinParams
-      QExecPlanUnresolved remoteRelField remoteSchemaInfo = unresolvedPlan
-
--- data JoinConfiguration = JoinConfiguration
-
-data JoinParams = JoinParams G.OperationType BatchInputs
+  deriving Show
+
+
+mkQuery :: G.OperationType -> BatchInputs -> QExecPlanUnresolved -> (Batch, VQ.RemoteTopField)
```
[the remark this answers is not in the document]
Comment (jberryman) [reply server/src-lib/Hasura/GraphQL/Execute.hs]: I changed it to JoinArguments, hope you're okay with that. I'm realizing I don't have the terminology to distinguish the different senses of e.g. "argument" i.e. our code deals both with the arguments to a field in a _query_ and also the values we _pass_ as arguments during execution

[the remark this answers is not in the document]
Comment (jberryman) [reply server/src-lib/Hasura/GraphQL/Execute.hs]: I'm not sure, feel free to rename to something that makes sense to you after seeing what it's become!

File: server/src-lib/Hasura/GraphQL/Execute/RemoteJoins.hs [hunk 17]
```diff
@@ -49,110 +55,129 @@ import           Hasura.Prelude
...
     , rrRelFieldPath  :: !RelFieldPath
+    -- ^ Splice path.
     , rrAlias         :: !G.Alias
+    -- ^ A possible alias (provided by client) for `rtrName .
+    -- rmfRemoteRelationship . rrRemoteField`, otherwise this will be defaulted
+    -- to that value.
     , rrAliasIndex    :: !Int
+    -- ^ index of selection set at 'rrRelFieldPath' at which this remote field should be spliced.
+    -- This is morally the last component of 'rrRelFieldPath'.
     , rrPhantomFields :: ![Text]
```
Comment (tirumaraiselvan) [inline server/src-lib/Hasura/GraphQL/Execute/RemoteJoins.hs]: I feel this should be `rrPhantomFields :: ![G.Name]`

File: server/src-lib/Hasura/GraphQL/Execute/RemoteJoins.hs [hunk 18]
```diff
@@ -49,110 +55,129 @@ import           Hasura.Prelude
...
+--   https://reasonablypolymorphic.com/blog/higher-kinded-data/ 
+--   https://hackage.haskell.org/package/barbies
+
+-- | This could live in "Hasura.Prelude" if we want to do this elsewhere.
+type family Param (p :: k) x
+-- DataKinds:
+data RRF_P = RRF_Tree | RRF_Splice
+-- Use the special parameter to determine whether to keep or drop fields (when
+-- used just as a splice):
+type instance Param 'RRF_Tree   a = a
```
Comment (tirumaraiselvan) [inline server/src-lib/Hasura/GraphQL/Execute/RemoteJoins.hs]: Very interesting. So if I understand this correctly at a high level:

If `RemoteRelBranch RRF_Tree` then we will get the `arguments` and `selSet` fields 
else if `RemoteRelBranch RRF_Splice` then we will get `arguments` and `selSet` to be their unit values?

File: server/src-lib/Hasura/GraphQL/Execute/RemoteJoins.hs [hunk 19]
```diff
@@ -49,110 +55,129 @@ import           Hasura.Prelude
...
-newtype RemoteRelKey =
-  RemoteRelKey Int
-  deriving (Eq, Ord, Show, Hashable)
-
-data RemoteRelField =
-  RemoteRelField
-    { rrRemoteField   :: !RemoteField
-    , rrArguments     :: !ArgsMap
-    , rrSelSet        :: !SelSet
+-- | When parameterised by 'RRF_Tree' this is like a 'VQ.Field' but with a path
```
Comment (tirumaraiselvan) [inline server/src-lib/Hasura/GraphQL/Execute/RemoteJoins.hs]: Maybe this is nitpicky but I don't like the name `RemoteRelBranch` :)

Should we call this `RemoteRelAnnField` to keep some semblance that this is basically a Field?

Also do we really need this kind of parameterisation? Is this for type-safety or for readability?

File: server/src-lib/Hasura/GraphQL/Execute/RemoteJoins.hs [hunk 20]
```diff
@@ -254,19 +274,17 @@ insertBatchResults accumResp Batch{..} remoteResp =
...
-                           [] ->
-                             Left $
-                               "no remote objects left for joining at " <> showHashO accumData
-                           (remoteDataVal:rest) -> do
-                             spliced <- spliceRemote remoteDataVal accumData
-                             pure
-                               (OJ.Object spliced : hasuraRowsSoFar , rest)
-                    One ->
-                      Left "Cardinality mismatch: found array in hasura value, but expected object"
+                  -- Since we've reached the end of the path in an Array, we
```
Comment (tirumaraiselvan) [inline server/src-lib/Hasura/GraphQL/Execute/RemoteJoins.hs]: Should we make `inValue` similar to `inHashMap` i.e. the first pattern match is on `RelFieldPath` rather than `currentValue`?

File: server/src-lib/Hasura/GraphQL/Execute/RemoteJoins.hs (hunk 18 above)
Comment (jberryman) [reply server/src-lib/Hasura/GraphQL/Execute/RemoteJoins.hs]: Yep that's right

File: server/src-lib/Hasura/GraphQL/Execute/RemoteJoins.hs (hunk 19 above)
Comment (jberryman) [reply server/src-lib/Hasura/GraphQL/Execute/RemoteJoins.hs]: Yeah the parameterisation is important (or once again making them two separate types); the two forms really are different things; the absence of `rrSelSet` means it's no longer a tree, and so definitely not `Field`-like.

But I'm not too happy with the names either.

Comment (jberryman) [reply server/src-lib/Hasura/GraphQL/Execute/RemoteJoins.hs]: I find it really difficult to figure out what a function is doing (or supposed to be doing) if the domain is larger than it should be; the type signature becomes meaningless or misleading, you can't `traceShow` inputs and expect to understand what's going on, etc. 

So it's for readability I guess. Operationally it's probably also not good to hold on to a bunch of unneeded data.

File: server/src-lib/Hasura/GraphQL/Execute/RemoteJoins.hs (hunk 18 above)
Comment (jberryman) [reply server/src-lib/Hasura/GraphQL/Execute/RemoteJoins.hs]: The `Param` trick makes this one step fancier from a basic parameterized type, since we can avoid special unwrapping of the special fields (e.g. `getConst`), and it lets us make the parameter a tag that expresses the semantics so that the entire parameterized type hopefully tells us a bit more than e.g. `RemoteRelBranch (Const ())` or something

File: server/src-lib/Hasura/GraphQL/Execute/RemoteJoins.hs (hunk 20 above)
Comment (jberryman) [reply server/src-lib/Hasura/GraphQL/Execute/RemoteJoins.hs]: hm not sure there's a way to do that

**Comment by jberryman** [comment]
Sorry I ran out of time/energy to think of a better name for `RemoteRelBranch` . What does the "Ann" in  `RemoteRelAnnField` mean? For now I added a bit more explanation and a TODO

File: server/src-lib/Hasura/GraphQL/Execute/RemoteJoins.hs (hunk 20 above)
Comment (tirumaraiselvan) [reply server/src-lib/Hasura/GraphQL/Execute/RemoteJoins.hs]: haha, I think it's like this for a reason then :)

**Comment by tirumaraiselvan** [comment]
> Sorry I ran out of time/energy to think of a better name for `RemoteRelBranch` . What does the "Ann" in `RemoteRelAnnField` mean? For now I added a bit more explanation and a TODO

`Ann` is just our internal abbreviation for annotation. It's fine. Let's do this later (if at all).
